pkg/diagnose/runner.go

```diff
@@ -17,8 +17,307 @@ import (
 	"github.com/fatih/color"
 )
 
-// RunAll runs all registered connectivity checks, output it in writer
-func RunAll(w io.Writer) error {
+func init() {
+	diagnosis.Register("connectivity-datadog-autodiscovery", diagnoseMetadataAutodiscoveryConnectivity)
+}
+
+// Overall running statistics
+type diangosisCounters struct {
+	totalCnt   int
+	successCnt int
+	failCnt    int
+	warningCnt int
+	errorCnt   int
+	skippedCnt int
+}
+
+// Output summary
+func outputSummary(w io.Writer, c diangosisCounters) {
+	fmt.Fprintf(w, "-------------------------\n  Total:%d", c.totalCnt)
+	if c.successCnt > 0 {
+		fmt.Fprintf(w, ", Success:%d", c.successCnt)
+	}
+	if c.failCnt > 0 {
+		fmt.Fprintf(w, ", Fail:%d", c.failCnt)
+	}
+	if c.warningCnt > 0 {
+		fmt.Fprintf(w, ", Warning:%d", c.warningCnt)
+	}
+	if c.errorCnt > 0 {
+		fmt.Fprintf(w, ", Error:%d", c.errorCnt)
+	}
+	fmt.Fprint(w, "\n")
+}
+
+func incrementCounters(r diagnosis.DiagnosisResult, c *diangosisCounters) {
+	c.totalCnt++
+
+	if r == diagnosis.DiagnosisNotEnable {
+		c.skippedCnt++
+	} else if r == diagnosis.DiagnosisSuccess {
+		c.successCnt++
+	} else if r == diagnosis.DiagnosisFail {
+		c.failCnt++
+	} else if r == diagnosis.DiagnosisWarning {
+		c.warningCnt++
+	} else { //if d.Result == diagnosis.DiagnosisUnexpectedError
+		c.errorCnt++
+	}
+}
+
+func getDiagnosisResultForOutput(r diagnosis.DiagnosisResult) string {
+	var result string
+	if r == diagnosis.DiagnosisSuccess {
+		result = color.GreenString("PASS")
+	} else if r == diagnosis.DiagnosisFail {
+		result = color.RedString("FAIL")
+	} else if r == diagnosis.DiagnosisWarning {
+		result = color.YellowString("WARNING")
+	} else { //if d.Result == diagnosis.DiagnosisUnexpectedError
+		result = color.HiRedString("UNEXPECTED ERROR")
+	}
+
+	return result
+}
```

Comment:
I have the same similar and small concern of coupling behavior and structure. Not particular strong and I do not mind changing it since it is a bit of a grey area in that regard I think